### PR TITLE
core: add attrs to request message

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
This is an example of adding Attrs (as an alternative to SessionContext).  Attrs are type safe, and make it easy to find out where a particular Key is used.  Attrs is mutable as there are several places where it its mutated.    Since we don't have the error prone checks to ensure return values are used, it would be risky to make Attrs immutable.

### CODE REVIEW

File: zuul-core/src/main/java/com/netflix/zuul/message/ZuulMessageImpl.java [hunk 1]
```diff
@@ -31,6 +32,8 @@
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
+import java.util.Objects;
+import org.w3c.dom.Attr;
```
Comment (artgon) [inline zuul-core/src/main/java/com/netflix/zuul/message/ZuulMessageImpl.java:36]: wrong import?

Comment (carl-mastrangelo) [reply zuul-core/src/main/java/com/netflix/zuul/message/ZuulMessageImpl.java]: fixed

**Review by artgon — APPROVED** [review]

**Comment by artgon** [comment]
I like it conceptually but I don't like the name "Attrs". For example, I think `SessionContext` is a lot more descriptive.

**Comment by carl-mastrangelo** [comment]
@artgon  I can make the accessor something like `callAttrs`.  I feel Session is the wrong terminology too (a session persists across multiple calls).   I am using the `Attrs` in other places so I don't think it would be correct to change the type name, just the variable name.

**Comment by artgon** [comment]
What about `RequestAttrs`?

**Comment by argha-c** [comment]
Yeah, given the usage, the name should allude to this being a request level attribute store. Avoid `RequestContext` since that has connotations based on prior usage in other frameworks/libs.

**Comment by argha-c** [comment]
Unrelated to this change but since we're on this class, https://github.com/Netflix/zuul/blob/500295cde056968db4156195d330662c0cc5ae67/zuul-core/src/main/java/com/netflix/zuul/Attrs.java#L65-L67
I think the type conversion should happen on line 67, ie. after the null check, not before.

**Comment by carl-mastrangelo** [comment]
I'm also hesitant to use the word `Request` because its shared for the request and the response, same as SessionContext.

**Comment by artgon** [comment]
So, we need a word that includes both Request and Response but isn't "Session" :)

**Comment by carl-mastrangelo** [comment]
I think `Call` encompasses both.

**Pull request closed without merging**